### PR TITLE
PKCE support for SSO

### PR DESCRIPTION
**Pull request by avdb13**
## Description

In continuation of https://github.com/LemmyNet/lemmy/pull/4881.

Implements PKCE support, in order to mitigate against the threat of authorization code interception attacks.

Background reading: https://www.oauth.com/oauth2-servers/pkce

### CODE REVIEW

File: crates/db_schema/src/schema.rs
Comment (flamingo-cant-draw) [inline crates/db_schema/src/schema.rs:634]: You're missing the needed migrations in `/migrations`. Run `diesel migration generate oauth_pkce` and add the needed SQL. [See here for more info](https://diesel.rs/guides/getting-started).

**Comment by Nutomic** [comment]
@privacyguard Could you have a look at this?

**Comment by avdb13** [comment]
Frontend PR: https://github.com/LemmyNet/lemmy-ui/pull/2806
How can I test my changes locally?

File: crates/api_crud/src/user/create.rs
Comment (avdb13) [reply crates/api_crud/src/user/create.rs:235]: What is the idiomatic way for input validation here?

File: crates/api_common/src/utils.rs
Comment (dessalines) [inline crates/api_common/src/utils.rs:1165]: There's no need to return the code verifier when you just passed it in. Just return `LemmyResult<()>`

File: crates/api_crud/src/user/create.rs
Comment (dessalines) [inline crates/api_crud/src/user/create.rs:237]: No need to use a map here, either do an `if let Some(..` or pass the option into your `check_code_verifier` function.

File: crates/api_common/src/utils.rs
Comment (dessalines) [inline crates/api_common/src/utils.rs:1165]: Also this should probably be renamed to `check_pkce_code`

File: crates/api_crud/src/user/create.rs
Comment (dessalines) [inline crates/api_crud/src/user/create.rs:540]: I'm usually against mut forms, but it makes sense here.

```rust
if let Some(pkce_...) = ... { 
  form.push("...)
}
```

File: crates/api_crud/src/user/create.rs
Comment (dessalines) [inline crates/api_crud/src/user/create.rs:554]: It'd probably be better to include this in the error message if its vital. Otherwise just remove this line. Ideally after you've tested this and it works, you can remove it.

File: migrations/2024-11-12-080606_oauth_pkce/up.sql
Comment (dessalines) [inline migrations/2024-11-12-080606_oauth_pkce/up.sql:1]: Since we've merged other DB changes since then, you'll probably need to rename this migration to the current date, or anything after the last one.

File: crates/api_common/src/utils.rs
Comment (Nutomic) [reply crates/api_common/src/utils.rs:1165]: This function is only used in a single file, so move it directly there and make it private.

File: migrations/2024-11-12-080606_oauth_pkce/up.sql
Comment (Nutomic) [reply migrations/2024-11-12-080606_oauth_pkce/up.sql:1]: Ah because there is already a newer migration on this same branch. Right it needs to be renamed then.

**Comment by avdb13** [comment]
Converting it back to draft for now, I tested it against gitlab but without success as their OIDC implementation seems to be too unreliable. Testing it with `django-oauth-toolkit` tonight.

**Comment by avdb13** [comment]
This should be about correct.

**Comment by avdb13** [comment]
Testing it seems to result in "Invalid login credentials" for some reason.

**Comment by avdb13** [comment]
CI doesn't pass because of `lemmy-js-client`. If you generate the types again it should work.

Thanks for the tutorial @privacyguard , I decided to use the same code with a few small changes.

**Comment by avdb13** [comment]
CI doesn't pass because of `lemmy-js-client`. If you generate the types again it should work.

Thanks for the tutorial @privacyguard , I decided to use the same code with a few small changes.

I have been thinking about opening another PR in order to support the nonce claim for OIDC providers, after this PR hopefully gets merged.

File: crates/api_common/src/oauth_provider.rs
Comment (Nutomic) [inline crates/api_common/src/oauth_provider.rs:79]: You need to add this to make ci pass:

```suggestion
  #[cfg_attr(feature = "full", ts(optional))]
  pub pkce_code_verifier: Option<String>,
```

File: crates/api_common/src/utils.rs
Comment (Nutomic) [inline crates/api_common/src/utils.rs:1182]: This is only used in create.rs so move it directly in that file, and make the function private.

**Comment by avdb13** [comment]
Not sure why federation tests fail. All issues have been addressed.

**Comment by dessalines** [comment]
Its not you, its a different intermittent CI issue we're facing.

BTW in the future, plz use regular commits and merges rather than rebasing. We can't see the specific changes because you rewrote the history and force-pushed it.

**Comment by avdb13** [comment]
> Its not you, its a different intermittent CI issue we're facing.

Demo to prove this is the case anyway: https://files.catbox.moe/pijbbh.mkv


> BTW in the future, plz use regular commits and merges rather than rebasing. We can't see the specific changes because you rewrote the history and force-pushed it.

Apologies, I figured that the changes were limited in scope.

**Review by Nutomic — APPROVED** [review]

**Comment by avdb13** [comment]
Do PRs require approval of all reviewers?

**Review by dessalines — APPROVED** [review]

**Pull request merged**
